File: apps/web/src/lib/api/student.ts
import { api } from "../apiClient";
import type {
    StudentProfile,
    Education,
    WorkExperience,
    Project,
    StudentSkill,
    Certification,
    Language,
    Gender,
} from "./types";

export type StudentProfileInput = {
    firstName: string;
    lastName?: string;
    phone?: string;
    city?: string;
    dob?: string; // ISO
    gender?: Gender;
    bio?: string;
};

export type EducationInput = {
    institute: string;
    degree: string;
    fieldOfStudy?: string;
    startYear: number;
    endYear?: number;
    grade?: string;
    current?: boolean;
};

export type ExperienceInput = {
    company: string;
    title: string;
    location?: string;
    startDate: string; // ISO
    endDate?: string; // ISO
    current?: boolean;
    description?: string;
};

export type ProjectInput = {
    title: string;
    link?: string;
    description?: string;
    startDate?: string;
    endDate?: string;
};

export type SkillInput = { name: string; level?: number };
export type CertificationInput = {
    name: string;
    issuer?: string;
    issueDate?: string;
    credentialUrl?: string;
};
export type LanguageInput = { name: string; proficiency?: number };

export const studentApi = {
    // profile
    get_me: () => api.get<{ profile: StudentProfile }>("/student/me"),
    create: (input: StudentProfileInput) =>
        api.post<{ profile: StudentProfile }>("/student/me", input),
    update: (input: Partial<StudentProfileInput>) =>
        api.patch<{ profile: StudentProfile }>("/student/me", input),
    get_public: (userId: string) =>
        api.get<{ profile: StudentProfile }>(`/student/${userId}`),

    // education
    add_education: (input: EducationInput) =>
        api.post<{ education: Education }>("/student/me/educations", input),
    update_education: (id: string, input: Partial<EducationInput>) =>
        api.patch<{ ok: true }>(`/student/me/educations/${id}`, input),
    remove_education: (id: string) =>
        api.delete<{ ok: true }>(`/student/me/educations/${id}`),

    // experience
    add_experience: (input: ExperienceInput) =>
        api.post<{ experience: WorkExperience }>(
            "/student/me/experiences",
            input,
        ),
    update_experience: (id: string, input: Partial<ExperienceInput>) =>
        api.patch<{ ok: true }>(`/student/me/experiences/${id}`, input),
    remove_experience: (id: string) =>
        api.delete<{ ok: true }>(`/student/me/experiences/${id}`),

    // project
    add_project: (input: ProjectInput) =>
        api.post<{ project: Project }>("/student/me/projects", input),
    update_project: (id: string, input: Partial<ProjectInput>) =>
        api.patch<{ ok: true }>(`/student/me/projects/${id}`, input),
    remove_project: (id: string) =>
        api.delete<{ ok: true }>(`/student/me/projects/${id}`),

    // skill
    add_skill: (input: SkillInput) =>
        api.post<{ skill: { id: string; name: string }; link: StudentSkill }>(
            "/student/me/skills",
            input,
        ),
    remove_skill: (skillId: string) =>
        api.delete<{ ok: true }>(`/student/me/skills/${skillId}`),

    // certification
    add_certification: (input: CertificationInput) =>
        api.post<{ certification: Certification }>(
            "/student/me/certifications",
            input,
        ),
    update_certification: (id: string, input: Partial<CertificationInput>) =>
        api.patch<{ ok: true }>(`/student/me/certifications/${id}`, input),
    remove_certification: (id: string) =>
        api.delete<{ ok: true }>(`/student/me/certifications/${id}`),

    // language
    add_language: (input: LanguageInput) =>
        api.post<{ language: Language }>("/student/me/languages", input),
    update_language: (id: string, input: Partial<LanguageInput>) =>
        api.patch<{ ok: true }>(`/student/me/languages/${id}`, input),
    remove_language: (id: string) =>
        api.delete<{ ok: true }>(`/student/me/languages/${id}`),
};

File: apps/server-py/app/schemas/student.py
from datetime import datetime

from pydantic import BaseModel, Field, HttpUrl

from app.db.enums import Gender


class StudentProfileCreate(BaseModel):
    firstName: str = Field(min_length=1)
    lastName: str | None = None
    phone: str | None = None
    city: str | None = None
    dob: datetime | None = None
    gender: Gender | None = None
    bio: str | None = None


class StudentProfileUpdate(BaseModel):
    firstName: str | None = Field(default=None, min_length=1)
    lastName: str | None = None
    phone: str | None = None
    city: str | None = None
    dob: datetime | None = None
    gender: Gender | None = None
    bio: str | None = None


class EducationCreate(BaseModel):
    institute: str = Field(min_length=1)
    degree: str = Field(min_length=1)
    fieldOfStudy: str | None = None
    startYear: int
    endYear: int | None = None
    grade: str | None = None
    current: bool | None = None


class EducationUpdate(BaseModel):
    institute: str | None = Field(default=None, min_length=1)
    degree: str | None = Field(default=None, min_length=1)
    fieldOfStudy: str | None = None
    startYear: int | None = None
    endYear: int | None = None
    grade: str | None = None
    current: bool | None = None


class ExperienceCreate(BaseModel):
    company: str = Field(min_length=1)
    title: str = Field(min_length=1)
    location: str | None = None
    startDate: datetime
    endDate: datetime | None = None
    current: bool | None = None
    description: str | None = None


class ExperienceUpdate(BaseModel):
    company: str | None = Field(default=None, min_length=1)
    title: str | None = Field(default=None, min_length=1)
    location: str | None = None
    startDate: datetime | None = None
    endDate: datetime | None = None
    current: bool | None = None
    description: str | None = None


class ProjectCreate(BaseModel):
    title: str = Field(min_length=1)
    link: HttpUrl | None = None
    description: str | None = None
    startDate: datetime | None = None
    endDate: datetime | None = None


class ProjectUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1)
    link: HttpUrl | None = None
    description: str | None = None
    startDate: datetime | None = None
    endDate: datetime | None = None


class SkillAdd(BaseModel):
    name: str = Field(min_length=1)
    level: int | None = Field(default=None, ge=1, le=5)


class CertificationCreate(BaseModel):
    name: str = Field(min_length=1)
    issuer: str | None = None
    issueDate: datetime | None = None
    credentialUrl: HttpUrl | None = None


class CertificationUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1)
    issuer: str | None = None
    issueDate: datetime | None = None
    credentialUrl: HttpUrl | None = None


class LanguageCreate(BaseModel):
    name: str = Field(min_length=1)
    proficiency: int | None = Field(default=None, ge=1, le=5)


class LanguageUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1)
    proficiency: int | None = Field(default=None, ge=1, le=5)
